service/src/main/java/com/blog/service/impls/PostServiceImpl.java

```diff
@@ -185,30 +184,27 @@ public void deleteTagInPost(Long postId, Long tagId) {
             throw new InternalServerException(deleteTagInPost);
     }
 
-    private void updatePostTags(RequestPostDto post, List<Tag> postTags) {
-        removeTagInPost(post, postTags);
+    private void updatePostTags(RequestPostDto post) {
+        removeTagInPost(post.getId());
         addTagsToPost(post);
     }
 
-    private void removeTagInPost(RequestPostDto post, List<Tag> postTags) {
-        postTags.forEach(tag -> {
-            if (post.getTags().stream().noneMatch(t -> t.equals(tag.getId()))) {
-                if (!postDao.deleteTagInPost(post.getId(), tag.getId()))
-                    throw new InternalServerException(deleteTagInPost);
-            }
-        });
+    private void removeTagInPost(Long postId) {
+        postDao.deleteAllTagsInPost(postId);
     }
 
     private void addTagsToPost(RequestPostDto post) {
-        post.getTags().forEach(tag -> {
-            if (!postDao.checkTagInPostById(post.getId(), tag)) {
-                if (!postDao.addTagToPost(post.getId(), tag))
-                    throw new InternalServerException(addTagToPost);
-            }
-        });
+        postDao.addTagsToPost(post.getId(), post.getTags());
     }
 
-    private void addDataInPosts(List<ResponsePostDto> posts) {
-        posts.forEach(post -> post.setTags(tagService.getAllTagsByPostId(post.getId())));
+    private void addTagsToPosts(List<ResponsePostDto> posts) {
+        Set<Long> postsId = new HashSet<>();
+        posts.forEach(responsePostDto -> postsId.add(responsePostDto.getId()));
+        List<TagDto> tagDtoList = tagService.getAllTagsByPostsId(postsId);
+        posts.forEach(post -> {
```

Comment:
Think about it
private List<> getPostTags(postId, List<> allTags)
allTags.stream().filter( { condition } ).collect(Collectors.toList())